src/librustc/infer/error_reporting/mod.rs

```diff
@@ -219,10 +221,30 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
             }
         };
         let (prefix, span) = match *region {
-            ty::ReEarlyBound(ref br) => (
-                format!("the lifetime {} as defined on", br.name),
-                self.sess.codemap().def_span(self.hir.span(node)),
-            ),
+            ty::ReEarlyBound(ref br) => {
+                let mut sp = cm.def_span(self.hir.span(node));
+                if let Some(generics) = self.hir.get_generics(scope) {
+                    for param in &generics.params {
+                        if param.name.name().as_str() == br.name.as_str() {
```

Comment:
This loop feels like something that occurs more often (e.g. in lowering code or other diagnostics). Maybe pull it out into a method on `Generics` and do a quick grep around the source to see if any other places could benefit from an `fn(&Generics, Name) -> Option<&Param>`

Reply:
I made a cursory look for other possible uses but I couldn't any. I'm sure I'll find some while working on something else down the line, though.